src/java.base/share/classes/java/lang/StringUTF16.java

```diff
@@ -301,13 +302,13 @@ public static byte[] compress(final int[] val, int off, final int count) {
                     // Try to make a latin1 string from the copy
                     if (getChar(utf16, ndx) <= 0xff &&
                             compress(utf16, 0, latin1, 0, count) == count) {
-                        return latin1;     // latin1 success
+                        return new String.Proto(latin1, LATIN1);
                     }
                 }
-                return utf16;
+                return new String.Proto(utf16, UTF16);
             }
         }
-        return latin1;     // Latin1 success
+        return new String.Proto(latin1, LATIN1);
     }
```

Comment:
Does the escape analysis succeed in these cases.
When I tried something similar, performance suffered until I used locals and had only a single `new Proto(...)`

Reply:
Good question, I'll check (tomorrow)

Reply:
No allocation increase in these:
```
Name                                                     (size) Cnt     Base     Error      Test     Error   Unit  Change
StringConstructor.newStringFromCodePointRangedMixedBegin      7  15   23,783 ±   1,165    22,701 ±   1,311  ns/op   1,05x (p = 0,016 )
  :gc.alloc.rate                                                    3214,236 ± 165,068  3369,949 ± 200,156 MB/sec   1,05x (p = 0,019 )
  :gc.alloc.rate.norm                                                 80,000 ±   0,000    80,000 ±   0,000   B/op   1,00x (p = 0,053 )
  :gc.count                                                          480,000             504,000           counts
  :gc.time                                                           364,000             393,000               ms
StringConstructor.newStringFromCodePointRangedMixedBegin     64  15   32,052 ±   0,247    32,194 ±   0,293  ns/op   1,00x (p = 0,137 )
  :gc.alloc.rate                                                    7378,931 ±  56,952  7346,573 ±  66,739 MB/sec   1,00x (p = 0,138 )
  :gc.alloc.rate.norm                                                248,000 ±   0,000   248,000 ±   0,000   B/op   1,00x (p = 0,000 )
  ```

Reply:
Thanks for confirming.
I think in my attempt, I had some compound conditions and fall-throughs.
The immediate return in separate control flows would make it easier for the escape analysis to see the allocation did not escape.